File: backend/app/schemas/spot_rental.py
"""Spot rental schemas for validation."""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.schemas.spot_rental_billing import SpotRentalBillingResponse


class SpotRentalBase(BaseModel):
    """Base spot rental schema."""

    location_name: str | None = Field(None, max_length=100)
    location_address: str | None = None
    check_in_date: date
    check_out_date: date | None = None
    nightly_rate: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    weekly_rate: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    monthly_rate: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    electric: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    water: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    waste: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    total_cost: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    amenities: str | None = None
    notes: str | None = None


class SpotRentalCreate(SpotRentalBase):
    """Schema for creating a spot rental."""

    pass


class SpotRentalUpdate(BaseModel):
    """Schema for updating a spot rental."""

    location_name: str | None = Field(None, max_length=100)
    location_address: str | None = None
    check_in_date: date | None = None
    check_out_date: date | None = None
    nightly_rate: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    weekly_rate: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    monthly_rate: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    electric: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    water: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    waste: Decimal | None = Field(None, ge=0, le=9999.99, decimal_places=2)
    total_cost: Decimal | None = Field(None, ge=0, le=99999.99, decimal_places=2)
    amenities: str | None = None
    notes: str | None = None


class SpotRentalResponse(SpotRentalBase):
    """Schema for spot rental response."""

    id: int
    vin: str
    created_at: datetime
    billings: list[SpotRentalBillingResponse] = []

    model_config = {"from_attributes": True}


class SpotRentalListResponse(BaseModel):
    """Schema for list of spot rentals."""

    spot_rentals: list[SpotRentalResponse]
    total: int

File: frontend/src/schemas/spotRental.ts
import { z } from 'zod'
import type { TFunction } from 'i18next'
import { makeDateSchema, makeNotesSchema } from './shared'

/**
 * Spot rental schema matching backend Pydantic validators.
 * See: backend/app/schemas/spot_rental.py
 *
 * CRITICAL: This schema fixes 8 missing isNaN validation bugs in SpotRentalForm
 *
 * Factory, not a constant — see the header of schemas/auth.ts for why.
 */

// Currency validators specific to spot rental limits - forms use valueAsNumber: true
const makeNightlyRateSchema = (t: TFunction) =>
  z
    .number()
    .min(0, t('common:validation.spotRental.nightlyRateNegative'))
    .max(9999.99, t('common:validation.spotRental.nightlyRateTooLarge'))

const makeLargeRateSchema = (t: TFunction) =>
  z
    .number()
    .min(0, t('common:validation.spotRental.rateNegative'))
    .max(99999.99, t('common:validation.spotRental.rateTooLarge'))

const makeUtilitySchema = (t: TFunction) =>
  z
    .number()
    .min(0, t('common:validation.spotRental.utilityNegative'))
    .max(9999.99, t('common:validation.spotRental.utilityTooLarge'))

// Optional versions - handle NaN from empty inputs
const makeOptionalNightlyRateSchema = (t: TFunction) =>
  makeNightlyRateSchema(t)
    .or(z.nan())
    .transform(val => isNaN(val) ? undefined : val)
    .optional()

const makeOptionalLargeRateSchema = (t: TFunction) =>
  makeLargeRateSchema(t)
    .or(z.nan())
    .transform(val => isNaN(val) ? undefined : val)
    .optional()

const makeOptionalUtilitySchema = (t: TFunction) =>
  makeUtilitySchema(t)
    .or(z.nan())
    .transform(val => isNaN(val) ? undefined : val)
    .optional()

export const makeSpotRentalSchema = (t: TFunction) =>
  z.object({
    location_name: z
      .string()
      .max(100, t('common:validation.spotRental.locationNameTooLong'))
      .optional(),
    location_address: z.string().optional(),
    check_in_date: makeDateSchema(t),
    check_out_date: z.string().optional(),
    nightly_rate: makeOptionalNightlyRateSchema(t),
    weekly_rate: makeOptionalLargeRateSchema(t),
    monthly_rate: makeOptionalLargeRateSchema(t),
    electric: makeOptionalUtilitySchema(t),
    water: makeOptionalUtilitySchema(t),
    waste: makeOptionalUtilitySchema(t),
    total_cost: makeOptionalLargeRateSchema(t),
    amenities: z.string().optional(),
    notes: makeNotesSchema(t).optional(),
  })

// Use z.output for Zod v4 compatibility with z.coerce fields
export type SpotRentalInput = z.input<ReturnType<typeof makeSpotRentalSchema>>
export type SpotRentalFormData = z.output<ReturnType<typeof makeSpotRentalSchema>>
